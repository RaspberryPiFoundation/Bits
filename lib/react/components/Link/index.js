var _jsxFileName = 'src/components/Link/index.js',
    _this = this;

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

import classnames from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import { Link as RouterLink } from 'react-router-dom';
// import './Link.scss'

var Link = function Link(_ref) {
  var className = _ref.className,
      to = _ref.to,
      props = _objectWithoutProperties(_ref, ['className', 'to']);

  var classes = classnames('c-link', className);

  return React.createElement(
    RouterLink,
    Object.assign({ className: classes, to: to }, props, {
      __source: {
        fileName: _jsxFileName,
        lineNumber: 11
      },
      __self: _this
    }),
    children
  );
};

Link.propTypes = {
  className: PropTypes.string,
  to: PropTypes.string.isRequired
};

export default Link;