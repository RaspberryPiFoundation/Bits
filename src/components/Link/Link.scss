@import '../../styles/component-helper';

.c-link {
  @include link;
}

.c-link--light-on-dark {
  @include link-light-on-dark;
}
