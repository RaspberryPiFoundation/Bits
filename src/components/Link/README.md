```jsx
import { MemoryRouter } from 'react-router-dom';
<MemoryRouter>
  <Link to="https://www.raspberrypi.org">Link text goes here</Link>
</MemoryRouter>
```

#### In HTML/CSS

```html static
<a class="c-link" href="https://www.raspberrypi.org">Link text goes here</a>
```
