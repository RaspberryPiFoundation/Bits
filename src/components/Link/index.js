import { Base } from 'react-iotacss'
import { Link as RouterLink } from 'react-router-dom'
import classnames from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'

import './Link.scss'

const Link = ({ children, className, excludeClassname, to, ...props }) => {
  const classNames = classnames(
    {
      'c-link': !excludeClassname,
    },
    className,
  )

  const anchorLink = (
    <Base className={classNames} href={to} tagName="a" {...props}>
      {children}
    </Base>
  )

  const isExternal = () => !/^https?:\/\//i.test(to)

  const routerLink = (
    <RouterLink className={classNames} to={to} {...props}>
      {children}
    </RouterLink>
  )

  if (isExternal()) {
    return anchorLink
  }

  return routerLink
}

Link.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  excludeClassname: PropTypes.bool,
  to: PropTypes.string.isRequired,
}

export default Link
