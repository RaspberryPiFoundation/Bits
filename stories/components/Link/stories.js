import React from 'react'
import { storiesOf } from '@storybook/react'

import { StaticRouter } from 'react-router-dom'

import Link from '../../../src/components/Link'

const context = {}

storiesOf('Link', module)
  .add('Internal link', () => (
    <StaticRouter context={context}>
      <Link to="/here">An internal link</Link>
    </StaticRouter>
  ))
  .add('Light on dark internal link', () => (
    <StaticRouter context={context}>
      <Link className="c-link--light-on-dark" to="/">
        Hello Link
      </Link>
    </StaticRouter>
  ))
  .add('external link', () => (
    <StaticRouter context={context}>
      <Link to="https://raspberrypi.org">An external link</Link>
    </StaticRouter>
  ))
  .add('Light on dark external link', () => (
    <StaticRouter context={context}>
      <Link className="c-link--light-on-dark" to="https://raspberrypi.org">
        Hello Link
      </Link>
    </StaticRouter>
  ))
