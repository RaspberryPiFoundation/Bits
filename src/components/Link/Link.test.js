import { mount, shallow } from 'enzyme'
import { Link as RouterLink } from 'react-router-dom'
import { MemoryRouter } from 'react-router-dom'
import Link from '../Link'
import React from 'react'

let internalLink

const children = 'Whatever'
const internalTo = '/path/to/wherever'
const externalTo = 'https://www.100yen.co.uk'

describe('<Link />', () => {
  describe('Renders...', () => {
    beforeEach(() => {
      internalLink = shallow(<Link to={internalTo}>{children}</Link>)
    })

    it('without crashing', () => {
      expect(internalLink.find('.c-link')).toBePresent()
    })

    it('matches its snapshot', () => {
      expect(internalLink).toMatchSnapshot()
    })

    it('renders a React Router Link component', () => {
      expect(internalLink.find(RouterLink)).toBePresent()
    })

    describe('external links', () => {
      it('renders an anchor link', () => {
        let externalLink = mount(
          <MemoryRouter>
            <Link to={externalTo}>{children}</Link>
          </MemoryRouter>,
        )
        expect(externalLink.html()).toEqual(
          '<a class="c-link " href="https://www.100yen.co.uk">Whatever</a>',
        )
      })
    })
  })

  describe('Modifiers', () => {
    describe('lightOnDark', () => {
      it('adds the correct className', () => {
        internalLink = shallow(
          <Link lightOnDark to={internalTo}>
            {children}
          </Link>,
        )

        expect(internalLink.find('.c-link')).toHaveClassName(
          'c-link--light-on-dark',
        )
      })
    })

    describe('excludeClassname', () => {
      it("doesn't have any classNames", () => {
        internalLink = shallow(
          <Link excludeClassName lightOnDark to={internalTo}>
            {children}
          </Link>,
        )

        expect(internalLink.find('a')).not.toHaveClassName(
          'c-link',
          'c-link--light-on-dark',
        )
      })
    })
  })
})
