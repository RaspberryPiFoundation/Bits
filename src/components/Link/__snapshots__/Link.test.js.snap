// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> Renders... matches its snapshot 1`] = `
<Base
  className="c-link"
  href="/path/to/wherever"
  tagName="a"
>
  Whatever
</Base>
`;
